### ISSUE TITLE
checkstyle false positive for RedundantModifier

### ISSUE THREAD

**Issue by hgschmie** [checkstyle/checkstyle#12216]
I have read check documentation: https://checkstyle.org/config_xxxxxx.html#NameOfAffectedCheck
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ cat YOUR_FILE.java
```

```java
class Foo {
    public static class Bar {
        public Bar() {}

        public static class Baz {
            public Baz() {}
        }
    }
}
```

```bash
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml YOUR_FILE.java

Foo.java:[3,9] (modifier) RedundantModifier: Redundant 'public' modifier.
Foo.java:[6,13] (modifier) RedundantModifier: Redundant 'public' modifier.
```
---

The class is a valid java class and no redundant modifiers are listed. 

---

This problem has existed forever (at least back to 8.20). Checkstyle believes that the "public" modifiers on the c'tors are redundant because the public classes are part of a package private class. But that is not true. If the c'tors are not public, then reflection can not invoke the c'tors.

**Comment by romani** [comment]
Please put in xx required information.
Please strictly follow issue template to provide all details we need

**Comment by delanym** [comment]
@romani I know its frustrating when people dont follow procedure, but the level of inflexibility here is really quite surprising. It feels more like dealing with a revenue service than maintainers of an open source software project.
If you have to post about not following the template all the time maybe you should consider being more lenient. We're all trying to help each other after all. 

This is a perfectly valid issue, and fairly obvious. The documentation https://checkstyle.sourceforge.io/config_modifier.html#RedundantModifier
incorrectly states that  `There is no violation in the following example, because removing public modifier from ProtectedInnerClass constructor will make this code not compiling: `. In fact, there are numerous violations.

**Comment by delanym** [comment]
I tried removing the token [CTOR_DEF](https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF) from the list of default tokens and found that the following tokens listed as "default" are unacceptable. [ANNOTATION_DEF](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ANNOTATION_DEF) ,[ RECORD_DEF](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF)

**Comment by delanym** [comment]
@hgschmie to invoke private constructors via reflection you need to use https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/reflect/AccessibleObject.html#setAccessible(boolean)
There some complication when it comes to the JPMS: https://stackoverflow.com/questions/63407349/java-illegal-access-operations-method-will-be-deprecated

Since its not possible for checkstyle to discover whether they would be used through reflection, perhaps there should be a further option to configure whether a public inner class > public constructor is considered redundant, in the case where you don't want your code using "setAccessible" for whatever reason?

**Comment by hgschmie** [comment]
setAccessible is going away, as you noticed. That said, I am not accessing a private c'tor. I am accessing a public c'tor on a nested public class which happen to be enclosed by a package private class. The VM does this just fine, only checkstyle considers the "public" modifier on the c'tor as redundant. checkstyle does not complain that the "public" modifier on the c'tor is redundant for a regular public class. The problem here is that it extends some scoping rule ("if the outer class is package private, then c'tors can only be package private or less visible") and does not "reset" that scoping rule when it encounters the nested public class.

This warning is valid for a c'tor on a package private class. It is not valid for a nested public class within a package private class. 

(and thank you for the kind words in https://github.com/checkstyle/checkstyle/issues/12216#issuecomment-1410041153. I have noticed a tendency with some open source projects that getting barked at for "not following some imaginary process" has become more important than "thank you for the issue report". It is a good way to drive away contributors)

**Comment by romani** [comment]
If someone can update or create new issue with all required details,  we can approve quicker and let someone in web provide a fix.

**Comment by rnveach** [comment]
https://checkstyle.org/config_modifier.html#RedundantModifier

Checkstyle Example:
````
$ cat TestClass.java
class Foo {
    public static class Bar {
        public Bar() {}

        public static class Baz {
            public Baz() {}
        }
    }
}
class ModifiedFoo {
    public static class Bar {
        Bar() {}

        public static class Baz {
            Baz() {}
        }
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <property name="haltOnException" value="false"/>

    <module name="TreeWalker">
<module name="RedundantModifier"/>
    </module>
</module>

$ java -jar checkstyle-10.7.0-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[WARN] TestClass.java:3:9: Redundant 'public' modifier. [RedundantModifier]
[WARN] TestClass.java:6:13: Redundant 'public' modifier. [RedundantModifier]
Audit done.
````

Runtime Example:
````
	public static void main(String... arguments) {
		workClass(Foo.Bar.class);
		System.out.println();
		workClass(Foo.Bar.Baz.class);
		System.out.println();
		workClass(ModifiedFoo.Bar.class);
		System.out.println();
		workClass(ModifiedFoo.Bar.Baz.class);
	}

	private static void workClass(Class<?> clzz) {
		System.out.println(clzz.getName());
		for (Constructor<?> constructor : clzz.getConstructors()) {
			System.out.println(constructor);

			try {
				constructor.newInstance();
			} catch (InstantiationException | IllegalAccessException | IllegalArgumentException
					| InvocationTargetException e) {
				e.printStackTrace();
			}
		}
	}
````
Execution:
````
Foo$Bar
public Foo$Bar()

Foo$Bar$Baz
public Foo$Bar$Baz()

ModifiedFoo$Bar

ModifiedFoo$Bar$Baz
````

As seen above, without `public` on the constructors, reflection cannot access the constructors and instantiate the class.
